resources/type-system/inference.md

```diff
@@ -315,45 +320,60 @@ described below with a typing context as computed above.
 The actual returned type of a function literal with a block body is computed as
 follows.  Let `T` be `Never` if every control path through the block exits the
 block without reaching the end of the block, as computed by the **definite
-completion** analysis specified elsewhere.  Let `T` be `Null` if any control
+completion** analysis specified elsewhere, or if the function is a generator
+function.
+Let `T` be `Null` if the function is a non-generator function and any control
 path reaches the end of the block without exiting the block, as computed by the
 **definite completion** analysis specified elsewhere.  Let `K` be the typing
 context for the function body as computed above from the imposed return type
 schema.
-  - For each `return e;` statement in the block, let `S` be the inferred type of
-    `e`, using the local type inference algorithm described below with typing
-    context `K`, and update `T` to be `UP(flatten(S), T)` if the enclosing
-    function is `async`, or `UP(S, T)` otherwise.
-  - For each `return;` statement in the block, update `T` to be `UP(Null, T)`.
+  - If the enclosing function is a non-`async` non-generator function,
+    for each `return e;` statement in the block, let `S` be the inferred type
+    of `e`, using the local type inference algorithm described below with typing
+    context `K`, and update `T` to be `UP(S, T)`.
+  - If the enclosing function is marekd `async`, for each `return e;` statement
+    in the block, let `S` be the inferred type of `e`, using the local type
+    inference algorithm described below with typing context `FutureOr<K>`,
```

Comment:
I don't think this is quite right.  You only want to use `FutureOr<K>` when `K` is not empty.  That is, the way that this was specified before, I think `() async { return e; }` would infer `e` using `_` as the context, whereas as written I think this implies that you use `FutureOr<_>` as the typing context.  I believe this can be observable.

Reply:
Oh wonders! :)

OK, so if `K` is not `_`, use `FutureOr<K>` as context type schema for `e`, otherwise use `_`.

Reply:
Is it a *problem* to use `FutureOr<_>`, or is it just a *difference*. (We probably don't want a difference either, just want to make sure I'm not missing something problematic.)

Reply:
I'm confused - I don't see where this happens.  The text seems to uniformly infer `return e` in a context of `FutureOr<K>`, which means when `K` is `_`, it will infer it using `FutureOr<_>` no?